src/hotspot/share/opto/vectornode.cpp

```diff
@@ -2164,6 +2164,98 @@ Node* OrVNode::Identity(PhaseGVN* phase) {
+    case BoolTest::ult:
+    case BoolTest::ugt: {
+      BasicType bt = vect_type()->element_basic_type();
+      // For float and double, we don't know if either comparison operand is a
+      // NaN, NaN {le|ge|lt|gt} anything is false, resulting in inconsistent
+      // results before and after negation.
+      return is_integral_type(bt);
+    }
+    default:
+      return false;
+  }
+}
+
+// This function transforms the following patterns:
+//
+// For integer types:
+// (XorV (VectorMaskCmp src1 src2 cond) (Replicate -1))
+//    => (VectorMaskCmp src1 src2 ncond)
+// (XorVMask (VectorMaskCmp src1 src2 cond) (MaskAll m1))
+//    => (VectorMaskCmp src1 src2 ncond)
+// cond can be eq, ne, le, ge, lt, gt, ule, uge, ult and ugt.
+// ncond is the negative comparison of cond.
+//
+// For float and double types:
+// (XorV (VectorMaskCast (VectorMaskCmp src1 src2 cond)) (Replicate -1))
+//    => (VectorMaskCast (VectorMaskCmp src1 src2 ncond))
+// (XorVMask (VectorMaskCast (VectorMaskCmp src1 src2 cond)) (MaskAll m1))
+//    => (VectorMaskCast (VectorMaskCmp src1 src2 ncond))
+// cond can be eq or ne.
+Node* XorVNode::Ideal_XorV_VectorMaskCmp(PhaseGVN* phase, bool can_reshape) {
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+  // Transformations for predicated IRs are not supported for now.
+  if (is_predicated_vector() || in1->is_predicated_vector() ||
+      in2->is_predicated_vector()) {
+    return nullptr;
+  }
+
+  // XorV/XorVMask is commutative, swap VectorMaskCmp/Op_VectorMaskCast to in1.
+  if (in2->Opcode() == Op_VectorMaskCmp ||
+      (in2->Opcode() == Op_VectorMaskCast && in2->in(1)->Opcode() == Op_VectorMaskCmp)) {
+    swap(in1, in2);
+  }
+
+  const TypeVect* vmcast_vt = nullptr;
+  if (in1->Opcode() == Op_VectorMaskCast && in1->outcnt() == 1 &&
+      in1->in(1)->Opcode() == Op_VectorMaskCmp) {
+    vmcast_vt = in1->as_Vector()->vect_type();
+    in1 = in1->in(1);
+  }
+  if (in2->Opcode() == Op_VectorMaskCast) {
+    in2 = in2->in(1);
+  }
+  if (in1->Opcode() != Op_VectorMaskCmp || in1->outcnt() > 1 ||
+      !((VectorMaskCmpNode*) in1)->predicate_can_be_inverted() ||
+      !VectorNode::is_all_ones_vector(in2)) {
+    return nullptr;
+  }
+
+  // This is the same with BoolTest::negate(), but we can't call it with a
+  // BoolTest object because the comparison may be unsigned comparison, but
+  // BoolTest doesn't support unsigned comparisons.
+  BoolTest::mask neg_cond =
+      (BoolTest::mask) (((VectorMaskCmpNode*) in1)->get_predicate() ^ 4);
+
+  ConINode* predicate_node = phase->intcon(neg_cond);
+  const TypeVect* vt = in1->as_Vector()->vect_type();
+  Node* vmcmp = new VectorMaskCmpNode(neg_cond, in1->in(1), in1->in(2),
+                                      predicate_node, vt);
+  if (vmcast_vt != nullptr) {
+    // We optimized out an VectorMaskCast, and in order to ensure type
+    // correctness, we need to regenerate one. VectorMaskCast will be encoded as
+    // empty for types with the same size.
```

Comment:
```suggestion
    // a no-op (identity function) for types with the same size.
```
Or what do you mean by "empty"? `TOP`? All zeros?